compiler/rustc_builtin_macros/src/autodiff.rs

```diff
@@ -0,0 +1,707 @@
+        }
+    };
+    let mut d_ret_ty = match d_ret_ty.kind.clone() {
+        TyKind::Tup(ref tys) => tys.clone(),
+        TyKind::Path(_, rustc_ast::Path { segments, .. }) => {
+            if segments.len() == 1 && segments[0].args.is_none() {
+                let id = vec![segments[0].ident];
+                let kind = TyKind::Path(None, ecx.path(span, id));
+                let ty = P(rustc_ast::Ty { kind, id: ast::DUMMY_NODE_ID, span, tokens: None });
+                thin_vec![ty]
+            } else {
+                panic!("Expected tuple or simple path return type");
+            }
+        }
+        _ => {
+            // We messed up construction of d_sig
+            panic!("Did not expect non-tuple ret ty: {:?}", d_ret_ty);
+        }
+    };
+    if is_fwd(x.mode) {
+        if x.ret_activity == DiffActivity::Dual {
+            assert!(d_ret_ty.len() == 2);
+            // both should be identical, by construction
+            let arg = d_ret_ty[0].kind.is_simple_path().unwrap();
+            let arg2 = d_ret_ty[1].kind.is_simple_path().unwrap();
+            assert!(arg == arg2);
+            let sl: Vec<Symbol> = vec![arg, Symbol::intern("default")];
+            let tmp = ecx.def_site_path(&sl);
+            let default_call_expr = ecx.expr_path(ecx.path(span, tmp));
+            let default_call_expr = ecx.expr_call(new_decl_span, default_call_expr, thin_vec![]);
+            exprs.push(default_call_expr);
+        }
+    } else {
+        assert!(is_rev(x.mode));
+
+        if primal_ret {
+            // We have extra handling above for the primal ret
+            d_ret_ty = d_ret_ty[1..].to_vec().into();
+        }
+
+        for arg in d_ret_ty.iter() {
+            let arg = arg.kind.is_simple_path().unwrap();
+            let sl: Vec<Symbol> = vec![arg, Symbol::intern("default")];
+            let tmp = ecx.def_site_path(&sl);
+            let default_call_expr = ecx.expr_path(ecx.path(span, tmp));
+            let default_call_expr = ecx.expr_call(new_decl_span, default_call_expr, thin_vec![]);
+            exprs.push(default_call_expr);
+        }
+    }
+
+    let ret: P<ast::Expr>;
+    if exprs.len() > 1 {
+        let ret_tuple: P<ast::Expr> = ecx.expr_tuple(span, exprs);
+        ret = ecx.expr_call(new_decl_span, blackbox_call_expr.clone(), thin_vec![ret_tuple]);
+    } else if exprs.len() == 1 {
+        let ret_scal = exprs.pop().unwrap();
+        ret = ecx.expr_call(new_decl_span, blackbox_call_expr.clone(), thin_vec![ret_scal]);
+    } else {
+        assert!(!d_sig.decl.output.has_ret());
+        // We don't have to match the return type.
+        return body;
+    }
+    assert!(d_sig.decl.output.has_ret());
+    body.stmts.push(ecx.stmt_expr(ret));
+
+    body
+}
+
+#[cfg(not(llvm_enzyme))]
+fn gen_primal_call(
+    ecx: &ExtCtxt<'_>,
+    span: Span,
+    primal: Ident,
+    idents: Vec<Ident>,
+) -> P<ast::Expr> {
+    let has_self = idents.len() > 0 && idents[0].name == kw::SelfLower;
```

Comment:
This doesn't cover all cases, and if I do it's longer and IMHO a bit less readable, since the two Some cases overlap significantly, whereas they are currently merged:
```
    2     match idents.split_first() {                                                                                     
    3         Some((first, rest)) if first.name == kw::SelfLower => { &Ident, &[Ident]                                                                                   
    4             let args: ThinVec<_> =                                                                                  
    5                 rest.iter().map(|arg| ecx.expr_path(ecx.path_ident(span, *arg))).collect(); &Ident                  
    6             let self_expr = ecx.expr_self(span); P<Expr>                                                                                
    7             ecx.expr_method_call(span, self_expr, primal, args.clone())                                                                 
    8         },                                                                                                                                                                  
H   9         Some(args) => {     ■■ if this is intentional, prefix it with an underscore: `_args` (&Ident, &[Ident])                                                             
   10             let args: ThinVec<_> =                                                                                                                                          
   11             idents.iter().map(|arg| ecx.expr_path(ecx.path_ident(span, *arg))).collect(); &Ident                                                                            
   12             let primal_call_expr = ecx.expr_path(ecx.path_ident(span, primal)); P<Expr>                                                                                     
   13             ecx.expr_call(span, primal_call_expr, args)                                                                                                                     
   14         },                                                                                                                                                                  
   15         None => {                                                                                                                                                           
   16             let args: ThinVec<_> =                                                                                                                                          
   17             idents.iter().map(|arg| ecx.expr_path(ecx.path_ident(span, *arg))).collect(); &Ident                                                                            
   18             let primal_call_expr = ecx.expr_path(ecx.path_ident(span, primal)); P<Expr>                                                                                     
   19             ecx.expr_call(span, primal_call_expr, args)
   20         },
   21     }
```
Is it then ok to leave it, or shall I still change it to a match?

Reply:
I think it's fine to leave as is, or maybe something like

```rs
if let Some(args @ (first, rest)) = idents.split_first() {
    if first.name == kw::SelfLower { ... } else { ... }
} else {
    // None case
};
```